### DISCUSSION TITLE
c.json returning empty string for undefined values

### DISCUSSION THREAD

**Discussion by marcosrjjunior** [honojs/hono#2338] [Q&A]
I'm mainly creating this ticket to understand a bit more about why this is happening and potentially flag an issue.

One of my functions was returning `undefined` instead of `false` and the `c.json` is forcing the response to be an "" (empty string) instead of throwing an error.
![Screenshot 2024-03-11 at 8 51 27 am](https://github.com/honojs/hono/assets/5287262/4ca77143-ce93-4643-824b-b5bf6a188368)
**Response:** ""

---

Reading the [json implementation here](https://github.com/honojs/hono/blob/4ca5f60def726731b3afbf3f6137f0a4b92d8b27/src/context.ts#L469) (hopefully that is the correct place).

I noticed that there possibly some extra things going on there:
- `JSON.stringify(object)`. Why do we need to run throguh a JSON.stringify before returning?
- Re-creating the response

Because of the issues mentioned above I replaced it to use the main [response](https://developer.mozilla.org/en-US/docs/Web/API/Response/json_static) instead.
![Screenshot 2024-03-11 at 8 51 36 am](https://github.com/honojs/hono/assets/5287262/d4c41bba-3362-47b0-8b9d-3e96411ceb1a)

**Comment by yusukebe** [comment]
Hi @marcosrjjunior !

Thanks for your comment. I think we can follow the `Response.json()` spec, but the results/content of `Response.json(undefined)` seem to be different between runtimes. Which runtime do you run it on?

By investigating, results are below:

* Cloudflare Workers - `"undefined"`
* Bun - `""`
* Deno - throw an error
* Node.js - throw an error

Ideally, `undefined` can't be serialized in JSON, so I think it should thows an error.

**Reply by yusukebe** [reply]
Indeed, according to MDN, it should throw an error if it's `undefined`.

https://developer.mozilla.org/en-US/docs/Web/API/Response/json_static#typeerror

**Comment by marcosrjjunior** [comment]
hey @yusukebe thanks for the quick reply.

I'm getting "" empty strings for both Nodejs and Bun, that is the issue.

I agree, it should thrown an error instead.

```
// ...
"hono": "^4.0.10",

"@hono/node-server": "^1.8.2",
```

**Reply by yusukebe** [reply]
@marcosrjjunior 

Thanks. I think the better way is to use `Response.json()` inside `c.json()`. When Hono was first developed, `Response.json()` was not implemented in all runtimes, so it is not used.

**Reply by marcosrjjunior** [reply]
that would be great @yusukebe!
leaving the c.json as close the the Response.json as possible without any other extra logic.
